pkg/deploy/strategy/rolling/rolling.go

```diff
@@ -0,0 +1,203 @@
+package rolling
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/golang/glog"
+
+	kapi "github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	kerrors "github.com/GoogleCloudPlatform/kubernetes/pkg/api/errors"
+	kclient "github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/runtime"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util/wait"
+
+	deployapi "github.com/openshift/origin/pkg/deploy/api"
+	"github.com/openshift/origin/pkg/deploy/strategy"
+	deployutil "github.com/openshift/origin/pkg/deploy/util"
+)
+
+// TODO: This should perhaps be made public upstream. See:
+// https://github.com/GoogleCloudPlatform/kubernetes/issues/7851
+const sourceIdAnnotation = "kubectl.kubernetes.io/update-source-id"
+
+// RollingDeploymentStrategy is a Strategy which implements rolling
+// deployments using the upstream Kubernetes RollingUpdater.
+//
+// Currently, there are some caveats:
+//
+// 1. When there is no existing prior deployment, deployment delegates to
+// another strategy.
+// 2. The interface to the RollingUpdater is not very clean.
+//
+// These caveats can be resolved with future upstream refactorings to
+// RollingUpdater[1][2].
+//
+// [1] https://github.com/GoogleCloudPlatform/kubernetes/pull/7183
+// [2] https://github.com/GoogleCloudPlatform/kubernetes/issues/7851
+type RollingDeploymentStrategy struct {
+	// initialStrategy is used when there are no prior deployments.
+	initialStrategy strategy.DeploymentStrategy
+	// client is used to deal with ReplicationControllers.
+	client kubectl.RollingUpdaterClient
+	// rollingUpdate knows how to perform a rolling update.
+	rollingUpdate func(config *kubectl.RollingUpdaterConfig) error
+	// codec is used to access the encoded config on a deployment.
+	codec runtime.Codec
+}
+
+// NewRollingDeploymentStrategy makes a new RollingDeploymentStrategy.
+func NewRollingDeploymentStrategy(namespace string, client kclient.Interface, codec runtime.Codec, initialStrategy strategy.DeploymentStrategy) *RollingDeploymentStrategy {
+	updaterClient := &rollingUpdaterClient{
+		GetReplicationControllerFn: func(namespace, name string) (*kapi.ReplicationController, error) {
+			return client.ReplicationControllers(namespace).Get(name)
+		},
+		UpdateReplicationControllerFn: func(namespace string, rc *kapi.ReplicationController) (*kapi.ReplicationController, error) {
+			return client.ReplicationControllers(namespace).Update(rc)
+		},
+		CreateReplicationControllerFn: func(namespace string, rc *kapi.ReplicationController) (*kapi.ReplicationController, error) {
+			return nil, fmt.Errorf("unexpected attempt to create deployment: %#v", rc)
```

Comment:
I guess that's fine.  Add a comment to that effect, and move ControllerHasDesiredReplicas up so it's above the guard functions (actual function, then guard functions with comment)